### PR TITLE
Private fontmap

### PR DESCRIPTION
**Pull request by jcupitt**
Use a private fontmap in `vips_text()` to avoid clashes with other pangocairo users in the same process.

See https://github.com/libvips/libvips/issues/3275

This is against 8.14, but maybe should be master? Or not. It needs a changelog note too, and I guess a version bump somewhere.

### CODE REVIEW

**Review by kleisauke — APPROVED** [review]
This regression was introduced in commit <https://github.com/libvips/libvips/commit/45a99238e26839584e2674cb72b2abdd6f711f4d>, a `GOnce` is indeed neater here.

LGTM!

**Comment by jcupitt** [comment]
Actually, thinking about it, we have to lock for the whole of text rendering :( I added a few more comments too.

**Comment by jcupitt** [comment]
@kleisauke you're the boss now, would you like to do final commit? Or I can if you prefer.

**Comment by kleisauke** [comment]
I still have to get used to the press the merge button, feel free to merge this one yourself. :)

**Comment by kleisauke** [comment]
BTW, I didn't notice a ["auto-merge" button](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/automatically-merging-a-pull-request#enabling-auto-merge), perhaps that would be useful [to enable within the repo settings](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/managing-auto-merge-for-pull-requests-in-your-repository#managing-auto-merge)? This would avoid having to manually wait for the CI before merging.

**Comment by jcupitt** [comment]
I think automerge is useful if you have rules like "approval from two reviewers required", but libvips hasn't been so formal in the past. 

I personally like a more freeform approach -- on work projects I've found the need to get multiple formal approvals on every change slows things down a lot and actually discourages discussion and experimentation (maybe that's the point). And nice software usually needs a lot of experimentation!

**Comment by kleisauke** [comment]
Ah, I see. I just noticed that the auto-merge option would only be available when there's a branch protection rule, see e.g.:
> Note: The option to enable auto-merge is shown only on pull requests that cannot be merged immediately. For example, when a branch protection rule enforces "Require pull request reviews before merging" or "Require status checks to pass before merging" and these conditions are not yet met. For more information, see "[Managing a branch protection rule](https://docs.github.com/en/github/administering-a-repository/defining-the-mergeability-of-pull-requests/managing-a-branch-protection-rule)."

Indeed, probably not useful for this project as you said.

File: libvips/create/text.c
Comment (jcupitt) [reply libvips/create/text.c:392]: It used to be the case that fontmap would not unref cleanly on some backends (win32? coretext? I forget), so it was important to keep them as long as possible and reuse them as much as possible. Perhaps this has been improved now? It was a while ago. 

If it hasn't been improved, then I think we should probably have a single fontmap we share between all instances of `text`.

Another alternative would be to scale text only in cairo rather than by changing dpi on the fontmap. This should be faster and more threadsafe, though I wonder if quality might be lower --- I would guess changing dpi is more likely to get the right hints sent down to the font rendering system.

Comment (kleisauke) [reply libvips/create/text.c:392]: Ah, I was unaware of that unref bug. Reusing the fontmap sounds OK in that case. I also wasn't sure how expensive the initial call to `pango_cairo_font_map_new()` would be, so perhaps a lock is faster than using a `GPrivate`.

Scaling via Cairo would likely reduce quality, which indeed should be handled by its font rendering backend (win32, Core Text, Fontconfig, etc.).

I think the issue with wxWidgets is that it does not use [`gtk_widget_set_font_map()`](https://docs.gtk.org/gtk4/method.Widget.set_font_map.html), so it would always end-up using the fontmap from `pango_cairo_font_map_get_default()`, see e.g.:
https://github.com/wxWidgets/wxWidgets/blob/662e23b432773515568224d00bf5baf269f5284a/src/gtk/window.cpp#L373
https://github.com/GNOME/gtk/blob/071c6c60c0784674bc52daf48fb0a94c19d28754/gtk/gtkwidget.c#L6400
https://github.com/GNOME/gtk/blob/071c6c60c0784674bc52daf48fb0a94c19d28754/gtk/gtkwidget.c#L6608

**Comment by kleisauke** [comment]
> It needs a changelog note too, and I guess a version bump somewhere.

FWIW, I started a changelog for 8.14.2 in PR #3290 and bumped the version in PR #3291.

File: libvips/create/text.c
Comment (kleisauke) [reply libvips/create/text.c:392]: IIUC (but correct me if I'm wrong), the `*_build` functions are rarely called concurrently on different threads (it's usually only called on the main thread). So, perhaps it does not matter so much whether you use a `GPrivate` or a lock here.

Let me know if I need to check the timing of calling `pango_cairo_font_map_new()`,  I suppose if that's an expensive one we'd better just reuse it with locking (i.e. this approach).

Comment (jcupitt) [reply libvips/create/text.c:392]: You can have several main threads, so `_build()` can run concurrently. I don't know how common this is though. 

The unref problems with `pango_cairo_font_map_new()` certainly used to be serious (large scale memory leaks with some backends), so I think it's probably best to just have one. Though that was a few years ago now, maybe these issues have been fixed? We'd need to do some testing to be sure.

The testing would be a PITA and the benefits of more threading in `text` seem small, so my 2p would be to just have one font map.

Comment (kleisauke) [reply libvips/create/text.c:392]: I agree, let's just have one font map.

**Review by kleisauke — APPROVED** [review]

**Comment by kleisauke** [comment]
Thanks! Merged to the master branch as well. Changelog note will be added via PR #3292.

**Pull request merged**
